design/design-doc-resource-configurations.md

```diff
@@ -0,0 +1,838 @@
+  # resourceRef, bindingPhase, and status conditions.
+  writeConnectionSecretToRef:
+    name: sql
+  classRef:
+    apiVersion: infrastructure.crossplane.io/v1alpha1
+    kind: ResourceConfigurationClass
+    name: private-mysql-server
+  resourceRef:
+    apiVersion: infrastructure.crossplane.io/v1alpha1
+    kind: ResourceConfiguration
+    name: default-sql-gdm0w
+---
+apiVersion: infrastructure.crossplane.io/v1alpha1
+kind: ResourceConfigurationClass
+metadata:
+  name: private-mysql-server
+spec:
+  # Each ResourceConfigurationClass defines how resources are dynamically
+  # provisioned for exactly one kind of resource claim; there is a one-to-one
+  # claim-kind-to-configuration-class-instance relationship.
+  claim:
+    apiVersion: database.example.org/v1alpha1
+    kind: MySQLInstance
+  # The array of primitive managed resources that will be dynamically
+  # provisioned to satisfy MySQLInstance claims that use this class. Each entry
+  # is an object containing a 'base' and an optional 'patch'. The base must be a
+  # valid managed resource (i.e. a defined custom resource that supports
+  # Crossplane's claim binding metadata). The patch specifies how the claim's
+  # fields may be used to extend or override the base's fields.
+  resources:
+  - base:
+      apiVersion: azure.crossplane.io/v1alpha3
+      kind: ResourceGroup
+      metadata:
+      spec:
+        # Required base fields (such as location) must specify a default value,
+        # even if it may be overridden by a patch.
+        location: West US
+        providerRef:
+          name: example
+        reclaimPolicy: Delete
+    # Each patch entry relates a claim field to a resource field, by fieldRef.
+    patch:
+    - fromClaimField: ".spec.region"
+      toResourceField: ".spec.forProvider.location"
+      # Patch entries may include an optional array of transform functions. This
+      # allows concepts (like region below) to be represented in a portable
+      # format at the claim level, and translated to the differing formats of
+      # each provider. For example "us-east" might map to "East US" in Azure and
+      # us-east1 in GCP. Note that the transform functions in these examples are
+      # purely illustrative; transform functions will be added as they are found
+      # to be useful.
+      transforms:
+      - type: map
+        map:
+          "us-west": "West US"
+          "us-east": "East US"
+  - base:
+      apiVersion: database.azure.crossplane.io/v1beta1
+      kind: MySQLServer
+      spec:
+        forProvider:
+          administratorLogin: myadmin
+          # A resource selector allows a resource dynamically provisioned using
+          # a particular resource configuration class to reference exactly one
+          # resource provisioned by the same class. In this case all
+          # MySQLServer resources provisioned using this class will select the
+          # ResourceGroup dynamically provisioned by the same class and claim.
```

Comment:
How would we handle a situation in which there were two `ResourceGroups` provisioned in a single configuration? It is my understanding that we will not be restricting configurations, even though it will likely be best practice to keep them as narrow as possible. However, a user could theoretically have a `ResourceConfigurationClass` that dynamically provisioned their whole organization's infrastructure, which could include many duplicates. I wonder if we could just add more attributes in the selector to say "match the resource in this configuration that satisfies all of these requirements".

Reply:
Keeping this comment for other reviewers, but I believe this is explained by the Cross Resource Referencing section below.

Reply:
> How would we handle a situation in which there were two ResourceGroups provisioned in a single configuration?

You'd label each resource group distinctly, and use both `matchConfiguration` and `matchLabels`, e.g.:

```yaml
resourceGroupNameSelector:
  matchConfiguration: true
  matchLabels:
    resource-group: a
```